File: bin/TerrainEditorData/LuaScripts/editheightui.lua
-- Elevation Height Editing Brush
require 'LuaScripts/editingbrush'

EditHeightUI=ScriptObject()

function EditHeightUI:Start()
	self.panel=ui:LoadLayout(cache:GetResource("XMLFile", "UI/TerrainEditHeightBrush.xml"))
	
	self:SubscribeToEvent("Pressed", "EditHeightUI:HandleButtonPress")
	self:SubscribeToEvent("SliderChanged", "EditHeightUI:HandleSliderChanged")
	
	self.panel.style=uiStyle
	ui.root:AddChild(self.panel)
	self.active=true
	
	self.brushpreview=Image(context)
	self.brushpreview:SetSize(64,64,3)
	self.brushtex=Texture2D:new(context)
	--self.brushtex:SetSize(0,0,0,TEXTURE_DYNAMIC)
	self.panel:GetChild("BrushPreview",true).texture=self.brushtex
	
	self.power,self.max,self.radius,self.hardness,self.usemask=self:GetBrushSettings()
	self:GenerateBrushPreview()
	
	local text=self.panel:GetChild("PowerText", true)
	if text then text.text=string.format("%.1f", self.power) end
	text=self.panel:GetChild("RadiusText", true)
	if text then text.text=tostring(math.floor(self.radius)) end
	text=self.panel:GetChild("MaxText", true)
	if text then text.text=string.format("%.1f", self.max) end
	text=self.panel:GetChild("HardnessText", true)
	if text then text.text=string.format("%.2f", self.hardness) end
	
	self.cursor=EditingBrush(scene_)
	self.cursor:BuildCursorMesh(self.radius)
	self.cursor:SetBrushPreview(self.brushtex)
	self.cursor:Show()
end

function EditHeightUI:GetBrushSettings()
	local power,max,radius,hardness=0,0,5,0.9
	local usemask0, usemask1, usemask2=false,false,false
	
	local slider
	slider=self.panel:GetChild("PowerSlider", true)
	if slider then power=(slider.value/slider.range)*4 end
	
	slider=self.panel:GetChild("MaxSlider", true)
	if slider then max=(slider.value/slider.range) end
	
	slider=self.panel:GetChild("RadiusSlider", true)
	if slider then radius=math.floor((slider.value/slider.range)*30) end
	
	slider=self.panel:GetChild("HardnessSlider", true)
	if slider then hardness=(slider.value/slider.range) end
	
	local button=self.panel:GetChild("Mask0Check", true)
	if button then usemask0=button.checked end
	
	button=self.panel:GetChild("Mask1Check", true)
	if button then usemask1=button.checked end
	
	button=self.panel:GetChild("Mask2Check", true)
	if button then usemask2=button.checked end
	
	return power,max,radius,math.min(1,hardness),usemask0, usemask1, usemask2
end

function EditHeightUI:GenerateBrushPreview()
	local hardness=0.5
	local slider=self.panel:GetChild("HardnessSlider", true)
	if slider then hardness=(slider.value/slider.range) end
	
	local w,h=self.brushpreview:GetWidth(), self.brushpreview:GetHeight()
	local rad=w/2
	local x,y
	for x=0,w-1,1 do
		for y=0,h-1,1 do
			local dx=x-w/2
			local dy=y-h/2
			local d=math.sqrt(dx*dx+dy*dy)
			--local i=(rad-d)/rad
			local i=(d-rad)/(hardness*rad-rad)
			i=math.max(0, math.min(1,i))
				
			self.brushpreview:SetPixel(x,y,Color(i*0.5,i*0.5,i*0.6))
		end
	end
	
	self.brushtex:SetData(self.brushpreview, false)
end

function EditHeightUI:GetBrushPreview()
	return self.brushtex
end


function EditHeightUI:Activate()
	self.panel.visible=true
	self.active=true
	self:GenerateBrushPreview(self.hardness)
	self.cursor:BuildCursorMesh(self.radius)
	self.cursor:Show()
	self.cursor:SetBrushPreview(self.brushtex)
end

function EditHeightUI:Deactivate()
	self.panel.visible=false
	self.active=false
	self.cursor:Hide()
end

function EditHeightUI:SetHeight(ht)
	local slider=self.panel:GetChild("MaxSlider", true)
	if slider then slider.value=ht*slider.range end
	self:GenerateBrushPreview()
end

function EditHeightUI:HandleButtonPress(eventType, eventData)

end

function EditHeightUI:HandleSliderChanged(eventType, eventData)
	local which=eventData["Element"]:GetPtr("UIElement")
	if which==nil then return end
	
	self.power, self.max, self.radius, self.hardness, self.usemask0, self.usemask1, self.usemask2=self:GetBrushSettings(self.panel)
	--self:BuildCursorMesh(self.radius)
	
	if which==self.panel:GetChild("PowerSlider", true) then
		local text=self.panel:GetChild("PowerText", true)
		if text then text.text=string.format("%.2f", self.power) end
	elseif which==self.panel:GetChild("RadiusSlider", true) then
		local text=self.panel:GetChild("RadiusText", true)
		if text then text.text=tostring(math.floor(self.radius)) end
		self.cursor:BuildCursorMesh(self.radius)
	elseif which==self.panel:GetChild("MaxSlider", true) then
		local text=self.panel:GetChild("MaxText", true)
		if text then text.text=string.format("%.2f", self.max) end
	elseif which==self.panel:GetChild("HardnessSlider", true) then
		local text=self.panel:GetChild("HardnessText", true)
		if text then text.text=string.format("%.3f", self.hardness) end
		self:GenerateBrushPreview(self.hardness)
	end
end

function EditHeightUI:Update(dt)
	if not self.active then return end
	
	
	local mousepos
	if input.mouseVisible then
		mousepos=input:GetMousePosition()
	else
		mousepos=ui:GetCursorPosition()
	end
	
	local ground=cam:GetScreenGround(mousepos.x, mousepos.y)
	
	if ground then 
		local world=Vector3(ground.x,0,ground.z)
		self.cursor:SetPosition(world)
		self.power, self.max, self.radius, self.hardness, self.usemask0, self.usemask1, self.usemask2=self:GetBrushSettings()
		
		if input:GetMouseButtonDown(MOUSEB_LEFT) and ui:GetElementAt(mousepos.x, mousepos.y)==nil then
			if input:GetQualifierDown(QUAL_CTRL) then
				local norm=WorldToNormalized(TerrainState.hmap,TerrainState.terrain,ground)
				local col=TerrainState.hmap:GetPixelBilinear(norm.x,1-norm.y)
				local ht=0
				if TerrainState.hmap.components==1 then ht=col.r
				else ht=col.r+col.g/256.0
				end
				print(ht)
				
				self:SetHeight(ht)
			else
				local gx,gz=ground.x,ground.z
				ApplyHeightBrush(TerrainState.terrain,TerrainState.hmap,TerrainState.mask,gx,gz,self.radius, self.max, self.power, self.hardness, self.usemask0, self.usemask1, self.usemask2, dt) TerrainState.terrain:ApplyHeightMap()
			end
		end
	end
	
	self.cursor:SetBrushCursorHeight()
end

File: Source/UI/editheightui.cpp
#include "editheightui.h"

#include "alphabrushselectorui.h"
#include "../terrainmaterialbuilder.h"
#include "../Components/editingcamera.h"

#include <Urho3D/Resource/ResourceCache.h>
#include <Urho3D/Resource/XMLFile.h>
#include <Urho3D/Graphics/Texture2D.h>
#include <Urho3D/UI/UIElement.h>
#include <Urho3D/UI/UI.h>
#include <Urho3D/Graphics/Graphics.h>
#include <Urho3D/UI/BorderImage.h>
#include <Urho3D/UI/Text.h>
#include <Urho3D/UI/Slider.h>
#include <Urho3D/UI/CheckBox.h>
#include <Urho3D/Input/Input.h>
#include <Urho3D/IO/Log.h>

#include "../format.h"


EditHeightUI::EditHeightUI(Context *context) : Object(context),
	terrainContext_(nullptr),
	materialBuilder_(nullptr),
	alphaSelector_(nullptr),
	camera_(nullptr),
	brushSettings_{0.5f, 0.5f, 1.0f, 0.0f},
	brushPreview_(context)
{
}

void EditHeightUI::Construct(TerrainContext *tc, TerrainMaterialBuilder *tmb, AlphaBrushSelectorUI *abs, EditingCamera *camera)
{
	terrainContext_=tc;
	materialBuilder_=tmb;
	alphaSelector_=abs;
	camera_=camera;

	auto cache=GetSubsystem<ResourceCache>();
	auto graphics=GetSubsystem<Graphics>();
	auto ui=GetSubsystem<UI>();

	XMLFile *style=cache->GetResource<XMLFile>("UI/DefaultStyle.xml");

	element_=ui->LoadLayout(cache->GetResource<XMLFile>("UI/TerrainEditHeightBrush.xml"), style);
	if(element_)
	{
		ui->GetRoot()->GetChild("Base",true)->AddChild(element_);
		brushPreview_.SetSize(64,64,3);
		brushPreviewTex_=SharedPtr<Texture2D>(new Texture2D(context_));
		element_->GetChildDynamicCast<BorderImage>("BrushPreview", true)->SetTexture(brushPreviewTex_);
		GenerateBrushPreview();
		SetBrushUIFields();
		
		element_->SetPosition(IntVector2(104, graphics->GetHeight()-element_->GetHeight()));

		SubscribeToEvent(element_->GetChild("PowerSlider", true), StringHash("SliderChanged"), URHO3D_HANDLER(EditHeightUI, HandleSliderChanged));
		SubscribeToEvent(element_->GetChild("RadiusSlider", true), StringHash("SliderChanged"), URHO3D_HANDLER(EditHeightUI, HandleSliderChanged));
		SubscribeToEvent(element_->GetChild("MaxSlider", true), StringHash("SliderChanged"), URHO3D_HANDLER(EditHeightUI, HandleSliderChanged));
		SubscribeToEvent(element_->GetChild("HardnessSlider", true), StringHash("SliderChanged"), URHO3D_HANDLER(EditHeightUI, HandleSliderChanged));
	}

	SubscribeToEvent(StringHash("Update"), URHO3D_HANDLER(EditHeightUI, HandleUpdate));
}

void EditHeightUI::SetVisible(bool v)
{
	if(element_) element_->SetVisible(v);
	if(alphaSelector_) alphaSelector_->SetVisible(v);

	if(!v) materialBuilder_->SetEditingCursor(-100000, -100000, brushSettings_.radius_*100.0f, brushSettings_.hardness_, camera_->GetYaw());
}

void EditHeightUI::HandleSliderChanged(StringHash eventType, VariantMap &eventData)
{
	GetBrushUIFields();

	Slider *slider;

	slider=element_->GetChildDynamicCast<Slider>("PowerSlider", true);
	slider->SetValue(brushSettings_.power_ * slider->GetRange());
	element_->GetChildDynamicCast<Text>("PowerText", true)->SetText("%.2f"_fmt(brushSettings_.power_));

	slider=element_->GetChildDynamicCast<Slider>("RadiusSlider", true);
	slider->SetValue(brushSettings_.radius_ * slider->GetRange());
	element_->GetChildDynamicCast<Text>("RadiusText", true)->SetText("%.2f"_fmt(brushSettings_.radius_));

	slider=element_->GetChildDynamicCast<Slider>("MaxSlider", true);
	slider->SetValue(brushSettings_.max_ * slider->GetRange());
	element_->GetChildDynamicCast<Text>("MaxText", true)->SetText("%.2f"_fmt(brushSettings_.max_));

	slider=element_->GetChildDynamicCast<Slider>("HardnessSlider", true);
	slider->SetValue(brushSettings_.hardness_ * slider->GetRange());
	element_->GetChildDynamicCast<Text>("HardnessText", true)->SetText("%.2f"_fmt(brushSettings_.hardness_));

	GenerateBrushPreview();
}

void EditHeightUI::GenerateBrushPreview()
{
	unsigned int w=brushPreview_.GetWidth();
	unsigned int h=brushPreview_.GetHeight();

	float rad=(float)w*0.5f;
	float hardness=std::min(0.99f, brushSettings_.hardness_);

	for(unsigned int y=0; y<h; ++y)
	{
		for(unsigned int x=0; x<w; ++x)
		{
			float dx=(float)x - (float)w*0.5f;
			float dy=(float)y - (float)h*0.5f;
			float d=std::sqrt(dx*dx+dy*dy);
			float i=(d-rad)/(hardness*rad-rad);
			brushPreview_.SetPixel(x,y,Color(i*0.5f, i*0.5f, i*0.5f));

		}
	}

	brushPreviewTex_->SetData(&brushPreview_, false);
}

void EditHeightUI::SetBrushUIFields()
{
	if(!element_) return;

	Slider *slider;

	slider=element_->GetChildDynamicCast<Slider>("PowerSlider", true);
	slider->SetValue(brushSettings_.power_ * slider->GetRange());
	element_->GetChildDynamicCast<Text>("PowerText", true)->SetText("%.2f"_fmt(brushSettings_.power_));

	slider=element_->GetChildDynamicCast<Slider>("RadiusSlider", true);
	slider->SetValue(brushSettings_.radius_ * slider->GetRange());
	element_->GetChildDynamicCast<Text>("RadiusText", true)->SetText("%.2f"_fmt(brushSettings_.radius_));

	slider=element_->GetChildDynamicCast<Slider>("MaxSlider", true);
	slider->SetValue(brushSettings_.max_ * slider->GetRange());
	element_->GetChildDynamicCast<Text>("MaxText", true)->SetText("%.2f"_fmt(brushSettings_.max_));

	slider=element_->GetChildDynamicCast<Slider>("HardnessSlider", true);
	slider->SetValue(brushSettings_.hardness_ * slider->GetRange());
	element_->GetChildDynamicCast<Text>("HardnessText", true)->SetText("%.2f"_fmt(brushSettings_.hardness_));


	CheckBox *button;
	button=element_->GetChildDynamicCast<CheckBox>("Mask0Check", true);
	button->SetChecked(maskSettings_.usemask0_);
	button=element_->GetChildDynamicCast<CheckBox>("Mask1Check", true);
	button->SetChecked(maskSettings_.usemask1_);
	button=element_->GetChildDynamicCast<CheckBox>("Mask2Check", true);
	button->SetChecked(maskSettings_.usemask2_);
}

void EditHeightUI::GetBrushUIFields()
{
	if(!element_) return;

	Slider *slider;

	slider=element_->GetChildDynamicCast<Slider>("PowerSlider", true);
	brushSettings_.power_ = slider->GetValue() / slider->GetRange();
	slider=element_->GetChildDynamicCast<Slider>("RadiusSlider", true);
	brushSettings_.radius_ = slider->GetValue() / slider->GetRange();
	slider=element_->GetChildDynamicCast<Slider>("MaxSlider", true);
	brushSettings_.max_ = slider->GetValue() / slider->GetRange();
	slider=element_->GetChildDynamicCast<Slider>("HardnessSlider", true);
	brushSettings_.hardness_ = std::min(0.99f, slider->GetValue() / slider->GetRange());

	CheckBox *button;
	button=element_->GetChildDynamicCast<CheckBox>("Mask0Check", true);
	maskSettings_.usemask0_=button->IsChecked();
	button=element_->GetChildDynamicCast<CheckBox>("Mask1Check", true);
	maskSettings_.usemask1_=button->IsChecked();
	button=element_->GetChildDynamicCast<CheckBox>("Mask2Check", true);
	maskSettings_.usemask2_=button->IsChecked();
}

void EditHeightUI::HandleUpdate(StringHash eventType, VariantMap &eventData)
{
	float dt=eventData["TimeStep"].GetFloat();
	if(!element_ || !element_->IsVisible()) return;

	auto input=GetSubsystem<Input>();
	auto ui=GetSubsystem<UI>();

	IntVector2 mousepos=input->GetMousePosition();

	Vector3 ground;
	bool s=camera_->GetMouseGround(ground);
	if(s)
	{
		materialBuilder_->SetEditingCursor(ground.x_, ground.z_, brushSettings_.radius_*100.0f, brushSettings_.hardness_, camera_->GetYaw());

		if(input->GetMouseButtonDown(MOUSEB_LEFT) && !ui->GetElementAt(mousepos.x_, mousepos.y_))
		{
			if(input->GetQualifierDown(QUAL_CTRL))
			{
				float ht=terrainContext_->GetHeightValue(ground);
				SetBrushMax(ht);
			}
			else
			{
				GetBrushUIFields();
				terrainContext_->ApplyHeightAlpha(ground.x_, ground.z_, dt, brushSettings_, maskSettings_, *(alphaSelector_->GetAlphaBrush()), -camera_->GetYaw()*3.14159265f/180.0f);
			}
		}
	}
}

void EditHeightUI::SetBrushMax(float ht)
{
	if(!element_) return;
	Slider *s=element_->GetChildDynamicCast<Slider>("MaxSlider", true);
	if(s)
	{
		s->SetValue(ht*s->GetRange());
		brushSettings_.max_ = ht;
		SetBrushUIFields();
	}
}
